### PR TITLE
Implementation of Kyma control loop

### PR DESCRIPTION
**Pull request by jeremyharisch**
This PR implements the control-loop for the KymaCR in the Watcher module. More details: https://github.com/kyma-project/kyma-watcher/issues/31

### CODE REVIEW

File: kcp/controllers/kyma_controller.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+	return modules
+}
+
+func (r *KymaReconciler) getWatcherCR(ctx context.Context,
+	module kyma.Module,
+	namespace string,
+) (*componentv1alpha1.Watcher, error) {
+	watcherCRName := fmt.Sprintf("%s-%s", module.Name, module.Channel)
+	watcherCR := &componentv1alpha1.Watcher{}
+	if err := r.Get(ctx, types.NamespacedName{Namespace: namespace, Name: watcherCRName}, watcherCR); err != nil {
```
Comment (adityabhatia) [inline kcp/controllers/kyma_controller.go:119]: I think you should skip if watcher CR was not found

File: kcp/controllers/kyma_controller.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	} else {
+		for key, value := range watcherConfigMap.Data {
+			if key == kymaCR.Name && value == kymaCR.Namespace {
+				// Kyma already exists in ConfigMap, nothing has to be done
+				return nil
+			}
+		}
+	}
+	// Kyma does not exists in ConfigMap
+	watcherConfigMap.Data[kymaCR.Name] = kymaCR.Namespace
```
Comment (adityabhatia) [inline kcp/controllers/kyma_controller.go:154]: this is not the intended use (end result) right? I thought config map will store a module list for all kymas?

**Review by adityabhatia — CHANGES_REQUESTED** [review]
minor comments, but looks great for the first iteration :fire:

File: kcp/controllers/kyma_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+	testConfigMap := NewTestConfigMap(configMapName)
+	testWatcherCR := NewTestWatcherCR(watcherName, map[string]string{})
+
+	SetupTestEnvironment(testKyma, testConfigMap, testWatcherCR)
+
+	It("should not throw an error in reconcile loop an ConfigMap should not be initialised nor updated", func() {
+		By("checking the data of the ConfigMap")
+		Eventually(GetConfigMapData(testConfigMap), 5*time.Second, interval).Should(BeEmpty())
+	})
+})
```
Comment (adityabhatia) [inline kcp/controllers/kyma_controller_test.go:55]: If you wish: I think these two tests can go to once spec, for simplicity

File: kcp/controllers/kyma_controller.go (hunk 2 above)
Comment (jeremyharisch) [reply kcp/controllers/kyma_controller.go:154]: Lets have a small discussion round regarding this. I had in mind, that we a one ConfigMap for each WatcherCR containing the KymaCRs the Watcher needs to take care of. Thus, we do not have too many racing conditions when updating the ConfigMap.

File: kcp/controllers/kyma_controller.go (hunk 1 above)
Comment (jeremyharisch) [reply kcp/controllers/kyma_controller.go:119]: Makes sense, will implement it in this way

File: kcp/controllers/kyma_controller_test.go (hunk 3 above)
Comment (jeremyharisch) [reply kcp/controllers/kyma_controller_test.go:55]: I can rewrite them, so they will fit in one spec

**Review by khlifi411 — APPROVED** [review]

**Pull request merged**